core/controller/src/main/scala/whisk/core/controller/Actions.scala

```diff
@@ -597,6 +650,89 @@ trait WhiskActionsApi extends WhiskCollectionAPI {
+        implicit transid: TransactionId): Future[Boolean] = {
+        // traverse all actions and "inline" all actions that are sequences
+        // keep track of all sequences to detect recursion
+        // first check that out of the box no more components than necessary
+        val future = if (components.size > whiskConfig.actionSequenceLimit.toInt) {
+            Future.failed(new TooManyActionsInSequence())
+        } else {
+            val resolvedSequence = WhiskAction.resolveAction(entityStore, sequence) // this assumes that entityStore is the same for actions and packages
+            val resolvedComponents = components map { c => WhiskAction.resolveAction(entityStore, c) }
+            val futureResComponents = Future.sequence(resolvedComponents)
+            futureResComponents flatMap { resComponents =>
+                // start inlining while keeping track of all sequences to check for recursion
+                resolvedSequence flatMap { sequence =>
+                    inlineComponentsAndCountAtomicActions(0, resComponents.toList, List(sequence))
+                }
+            }
+        }
+
+        future recoverWith {
+            case _: TooManyActionsInSequence =>
+                Future failed RejectRequest(BadRequest, "too many actions in sequence")
+            case _: SequenceWithRecursion =>
+                Future failed RejectRequest(BadRequest, "recursion detected in sequence")
+            case _: NoDocumentException =>
+                Future failed RejectRequest(BadRequest, "action not found")
+        }
+    }
+
+    /*
+     * helper function that traverses the actions to inline and counts atomic actions while keeping track of sequences
+     * @param atomicActionsCnt the atomic actions seen so far
+     * @param actionsToInline the actions to inline
+     * @param sequences the sequences encountered during inlining
+     * @return a successful future in case the conditions were met or a failed fugure with either too many actions or recursion found
+     */
+    private def inlineComponentsAndCountAtomicActions(atomicActionsCnt: Int, actionsToInline: List[FullyQualifiedEntityName], sequences: List[FullyQualifiedEntityName])(
+        implicit transid: TransactionId): Future[Boolean] = {
+        if (atomicActionsCnt > whiskConfig.actionSequenceLimit.toInt)
+            Future.failed(new TooManyActionsInSequence())
+        else {
+            actionsToInline match {
+                case action :: restActions =>
+                    val docid = action.toDocId
+                    WhiskAction.get(entityStore, docid) flatMap { act =>
+                        act.exec match {
+                            case seq: SequenceExec =>
+                                // this is actually a sequence, check already traversed sequences
+                                if (sequences.contains(action)) {
```

Comment:
this check was incorrect and generally this implementation did not do what was intended. this was corrected in a later commit.